src/BenchmarksApps/SslStream/Server/Program.cs

```diff
@@ -0,0 +1,177 @@
+            while (!ctrlC.IsCancellationRequested)
+            {
+                var clientSock = await sock.AcceptAsync(ctrlC.Token).ConfigureAwait(false);
+                _ = Task.Run(() => ProcessClient(clientSock, options, sslOptions, ctrlC.Token));
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Expected, just return
+        }
+
+        Log("Exitting...");
+    }
+
+    static async Task ReadWriteScenario(SslStream stream, ServerOptions options, CancellationToken cancellationToken)
+    {
+        static async Task WritingTask(SslStream stream, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (bufferSize == 0)
+            {
+                return;
+            }
+
+            var sendBuffer = new byte[bufferSize];
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    await stream.WriteAsync(sendBuffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected, just return
+            }
+        }
+
+        byte[] recvBuffer = new byte[options.ReceiveBufferSize];
+
+        CancellationTokenSource cts = new CancellationTokenSource();
+        using var reg = cancellationToken.Register(() => cts.Cancel());
+
+        var writeTask = Task.Run(() => WritingTask(stream, options.SendBufferSize, cts.Token));
+
+        var recvBufer = new byte[options.ReceiveBufferSize];
+
+        while (true)
+        {
+            if (await stream.ReadAsync(recvBufer, cts.Token).ConfigureAwait(false) == 0)
+            {
+                cts.Cancel();
+                break;
+            }
+        }
+
+        await writeTask.ConfigureAwait(false);
+    }
+
+    static async Task ProcessClient(Socket socket, ServerOptions options, SslServerAuthenticationOptions sslOptions, CancellationToken cancellationToken)
+    {
+        try
+        {
+            using var stream = await EstablishSslStreamAsync(socket, sslOptions, cancellationToken).ConfigureAwait(false);
+
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("Handshake"u8))
+            {
+                // done everything
+                return;
+            }
+            if (stream.NegotiatedApplicationProtocol.Protocol.Span.SequenceEqual("ReadWrite"u8))
+            {
+                await ReadWriteScenario(stream, options, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                Log($"Negotiated unknown protocol: {stream.NegotiatedApplicationProtocol}");
+            }
+        }
+        catch (AuthenticationException e)
+        {
+            Log($"Authentication failed: {e}");
+        }
+        catch (IOException e) when (e.InnerException is SocketException)
+        {
+            // client closed the connection, ignore this
```

Comment:
I think it will be better to count and report the number of errors (maybe even separately per kind). I believe, currently we've made the client to always finish "gracefully" in normal circumstances 🤔 am I missing something?

Also, even if we decide to ignore the last socket exception -- it should be the _last_ exception _only_ -- exceptions in the middle are wrong and should be reported to metrics.

_Just logging_ is not enough -- when the benchmark completes successfully (i.e. there were no unhandled exceptions), you normally don't see server's (and client's) logs, only the metrics. (esp. given that the end-goal is to add this as a power bi graph)

Reply:
Changed to report number of errors.